docs/src/references/agent/operation-workflow.md

```diff
@@ -0,0 +1,259 @@
+  - Having no other terminal states is important so any command requester (as the mapper)
+    can await the termination of a command in a systematic manner *even* if the workflow is updated.
+- A workflow API should define *no-op* states with no pre-defined actions
+  and which sole purpose is to give an agent developer the opportunity to *insert* its own logic and extra steps.
+  - A *no-op* state is a state which has a single direct transition to the next state.
+  - As an example, having a **downloaded** *no-op* state, with a direct transition to an **install** state which uses the downloaded file,
+    lets an agent operator override the automatic transition, say to check the downloaded file *before* moving the **install** state.
+  - The **init** state should be a *no-op* state.
+
+### Operation Implementation
+
+A workflow implementation for a specific operation must implement the actions specified for each non *no-op* state.
+This implementation has some freedom, notably to implement extra checks and actions but also to add new states.
+
+However, there are some rules and best practices.
+
+- All the state messages must be published as retained with QoS 1.
+- A workflow implementation should not reject a state message payload with unknown fields.
+  - It's also important to keep these unknown fields in the following states.
+  - This is important as we want to *extend* the workflow of an operation.
+    A software component added by the user might need these *extra* fields the plugin is not aware of.
+- A workflow implementation must not react on *no-op* states nor terminal states.
+  - The transition from a *no-op* state must be handled either by thin-edge as direct transition
+    or overridden by the user with domain-specific checks and actions.
+  - The terminal states, a.k.a **successful** and **failed**, are owned by the process which created the **init** state (in practice, the mapper).
+    Only this process should clear the retained message state for an operation instance by sending an empty payload on command's topic.
+
+### Workflow Overriding
+
+Thin-edge provides a mechanism to override, extend and combine workflows.
+
+This mechanism is provided by the **tedge-agent** which gathers a set of user-defined workflows
+and combined them with the builtin workflows implemented by the agent itself.
+
+Each workflow is defined using a TOML file stored in `/etc/tedge/operations`. Each specifies:
+- the command name that it should trigger on
+  such as `firmware_update` or `restart`- the list of states
+- for each state:
+  - the state name as defined by the operation API
+  - the set of states which can be an outcome for this state actions
+  - possible extra instructions on how to process the command at this stage, e.g.
+    - run a script
+
+```toml title="file: firmware_update_example.toml"
+operation = "firmware_update"
+
+[init]
+  script = "/usr/bin/firmware_handler.sh plan"
+  next = ["scheduled", "failed"]
```

Comment:
It may be obvious from these examples, but it might be better to explicitly document that the first value is the target state on a successful state action and the second one is for failure.

Reply:
The order of values does not mean anything at this point in time.

Reply:
I'm missing something here then. My understanding was that, if the script exits with non-zero exit code, without any JSON output with explicit `status`, then the workflow moves to the first state mentioned in the array. In the case of a script execution failure, it moves to the second. Isn't that the case?

Reply:
That isn't implemented yet, and we are still deciding on what to do there. This will be done in a follow up PR

Reply:
For now, this list is mostly documentation.